### PR TITLE
[ADD] New module 'mrp_production_hierarchy'

### PR DESCRIPTION
**Pull request by sebalix**
Production Orders Hierarchy
===========================

This module allows to view the hierarchy of generated production orders
when several levels of bill of materials are involved.

Usage
=====

Create a new production order (Manufacture / Operations / Manufacturing Orders)
with a multi-levels nomenclatured product, and related products configured with
'Manufacture' and 'Make To Order' routes.

Once saved, click on the 'Hierarchy' button at right.

Known issues / Roadmap
======================

* Only new production orders will benefit from this feature, it is not compatible with existing orders.

### CODE REVIEW

**Comment by pedrobaeza** [comment]
Maybe a screenshot in the README will help to understand and showcase how this works.

**Comment by sebalix** [comment]
@pedrobaeza there are already 3 screenshots?

**Comment by pedrobaeza** [comment]
Sorry, I didn't see them!

**Comment by JordiBForgeFlow** [comment]
@sebalix seems that you just show the first level of child, not the full hierarchy. Is that correct?

**Comment by sebalix** [comment]
@jbeficent exactly, the parent MO is not shown here, but it is feasible if you prefer?
EDIT : except the parent MO, the full hierarchy is displayed, it's just that the BoM used (standard demo data) for the screenshot has only 1 level...

**Comment by JordiBForgeFlow** [comment]
Ok! Thx

**Comment by sebalix** [comment]
Hierarchy treeview now includes the parent MO + Screenshots updated (the last one now shows a real hierarchy).

**Comment by JordiBForgeFlow** [comment]
Yes, I see 
image

**Comment by sebalix** [comment]
The module now displays the root and parent manufacturing order on a child's form.
I'm done with this PR, ready to be reviewed.

File: mrp_production_hierarchy/models/mrp_production.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    parent_id = fields.Many2one(
+        'mrp.production', u"Parent order",
+        index=True, ondelete='restrict', readonly=True)
+    child_ids = fields.One2many(
+        'mrp.production', 'parent_id', u"Child orders")
+    parent_left = fields.Integer('Left Parent', index=True)
+    parent_right = fields.Integer('Right Parent', index=True)
+
+    @api.multi
+    def _generate_moves(self):
```
Comment (JordiBForgeFlow) [inline mrp_production_hierarchy/models/mrp_production.py]: This is an api.multi. So self is likely to be a recordset, and in the  original method it iterates on that recordset.

IMHO you should iterate through the recordset,like:

for mo in self:
   if not mo.env.context.get('root_mrp_production_id'):
            mo = mo.with_context(root_mrp_production_id=mo.id)
   mo = mo.with_context(parent_mrp_production_id=mo.id)
   return super(MrpProduction, mo)._generate_moves()

Comment (sebalix) [reply mrp_production_hierarchy/models/mrp_production.py]: Thanks, fixed.

**Comment by sebalix** [comment]
We are experiencing issues with the `_parent_store =  True` setting on `mrp.production` when creating manufacturing orders concurrently: https://github.com/odoo/odoo/issues/14493
Solution by Odoo on 10.0 (related to `stock.quant.package` only): https://github.com/odoo/odoo/pull/14694
The commit b2e47e27ac497e189cfc8a9841d9b05581796d46 applies the same fix.

The `child_of` performance is related to the depth of the tree with `_parent_store = False`, and we are performing `child_of` search on MO trees with 400-500 children on 7-8 levels of depth without any problem.

**Comment by ovnicraft** [comment]
Here tested with 7 levels w/o problems.

**Review by LoisRForgeFlow — APPROVED** [review]
Functional review :+1:

**Comment by remihb** [comment]
LGTM ;)
Module in production for 3 months, no issues were noted.
Need merge ?

**Comment by LoisRForgeFlow** [comment]
@remihb next time do the review using github feature, so we can count your thumb up :)

/ocabot merge

**Comment by LoisRForgeFlow** [comment]
Runbot issue has been fixed here: https://github.com/OCA/manufacture/pull/401

/ocabot merge

**Pull request merged**